bevyengine/bevy
# Experimental viewports

This is a first (very incomplete) attempt at attaching Viewports (rather than Windows) to Cameras. I'm putting this up to serve as a basis for further discussion.

**Why would this be useful**
- Offscreen/headless rendering to a texture doesn't use a window
- It allows dividing a window/texture into different viewports, each with its own camera

**How to test this**
Just run any example, for instance:
```cargo run --example 3d_scene```
Also make sure to watch the console output as you resize the window.

**How is this implemented right now**
- There's a ViewportNode (should probably be renamed to LayoutNode or similar) connected to the target texture
- On every update, the node checks if the texture was resized
- If it was, it recomputes the viewport sizes, and sends corresponding ViewportResized events
- The camera system listens for these events and updates its own data structures

**What's missing to make this actually useful**
- There's no notion of Res<Viewports> yet, or even a working viewport identification. The default viewport is used all the time.
- Right now the ViewportNode only supports a single "Full" layout. A "split" layout for instance would create two Viewports.
- There's no actual support for rendering to a viewport in bevy yet (and I suspect that even wgpu only supports defining a single viewport at a time).

**Possible effects in performance**

First of all, the ViewportNode doesn't know when the render texture is resized, and checks the corresponding TextureDescriptor for changes in every update. It's fast and probably cheap, but not sure how it would compare to listening for events.

Second, there's now a visible latency between the window being resized and the camera projection being updated. Before both could be done in parallel, but now the projection is calculated AFTER the render texture is actually recreated.
